FAB-17890 Ch.Part.API: allow registrar to list a single channel

tock-ibm opened the pull request:
Implement the ChannelInfo(..) method in the registrar that looks up
a channel in the chains map and reports extended information on its
status. For this end we have to introduce the follower.Chain and
augment existing cluster type chains with means of reporting their
status.

- Introduce the skeleton a new type of consensus.Chain
  implementation: follower.Chain. This will be created and run when
  the orderer is required to follow the cluster and pull blocks from
  other orderers.

- The plan (for future commits) is for the follower.Chain to trigger
  the creation of an etcdraft.Chain when it discovers the orderer was
  added to the cluster, and vise versa; the etcdraft.Chain will replace
  itself with a follower.Chain when the orderer is removed from the
  cluster.

- Introduce a new interface that cluster-type chains implement, that
  allows them to report their relation to the cluster and their status.
  This is done because the registrar is not aware of the exact type of
  the chains it is keeping. The registrar cannot reflect on the type
  as well, as this will cause an import cycle (due to the etcdraft
  package importing multichannel).

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: Ia454f47f04a8ba3dcd76886a5919d1c734c01015

#### Type of change

- New feature

#### Related issues

Task: FAB-17980
Story: FAB-17824
Epic: FAB-17712

ghaskins commented on .travis.yml:
```diff
@@ -32,14 +32,12 @@ install:
  - cd /$HOME/gopath/src/github.com/hyperledger/$REPO_NAME/peer
  - sudo rm -rf /var/hyperledger/ && sudo mkdir /var/hyperledger/ && sudo chown $USER:$USER /var/hyperledger
- - go build && echo " STARTING PEER PROCESS "
+ - make peer && echo " STARTING PEER PROCESS "
```
I just noticed this isn't going to work.  I need to cd to the TLD, do the make, and then cd into peer.

ghaskins commented on docs/API/SandboxSetup.md:
```diff
@@ -23,9 +23,9 @@ From your command line terminal, move to the `devenv` subdirectory of your works
-    ./membersrvc
+    cd $GOPATH/src/github.com/hyperledger/fabric
+    make membersrvc
+    (cd menbersrvc; ./membersrvc)
```
typo in cd command

jyellick commented:
@ghaskins This looks like a great idea to me.  You said you chose make because it is ubiquitous and well understood.  Did you consider any other tools like Gradle (a tool I know @jeffgarratt is fond of)? I'm by no means an expert on build tools, just curious.

ghaskins commented:
@jyellick I didn't look at Gradle primarily b/c I am not familiar with it.  Of all the build tools, I am by far the most experienced with Make and to a lesser extent Maven and some of its derivatives like Leiningen. I figured there was probably some interesting projects out there, but my inclination was keep-it-simple so I went with the tried and true.  I am open to other ideas if there is a strong opinion that the assumed lower ubiquity/familiarity are worth the trade offs.

bcbrock commented:
I like the idea of having a Makefile - I was about to suggest this myself.  I have a similar Makefile in my busywork project which I will contrast with this proposed one.

First, I don't think the dependencies are working correctly. I would expect that if I say `make peer` that the peer would be rebuilt if any of the Go files changed. This does not happen for me  - I'm not sure what it would take to get the peer to be rebuilt here.

I have a target specifically to create the peer and membersrvc Docker images. In yours this is only a side-effect of running behave. (busywork also needs these images to run its Docker-compose networks).

The unit tests need to have a local peer running, don't they? That should be part of the make target I think. And the behave tests *can't* have a peer running, which should also be part of the make target.

For consistency in testing I have defined a lot of system-wide cleanup targets:
```
############################################################################
# "Clean" routines

.PHONY: docker_chaincode_clean docker_ips \
	docker_fabric_clean docker_ps_clean docker_reset \
	fabric_clean fabric_reset

docker_chaincode_clean: docker_ps_clean
	-docker images | egrep '^dev-|^<none>' | \
		awk '{print $$3}' | xargs docker rmi -f

docker_ips:
	docker ps -a -q | xargs -n 1 docker_ip	

docker_fabric_clean: docker_chaincode_clean
	-docker rmi -f hyperledger-peer membersrvc

docker_ps_clean:
	-docker ps -a -q | xargs docker rm -f

docker_reset:
	sudo service docker stop
	sudo rm -rf /var/lib/docker
	sudo service docker start

fabric_clean:
	-pkill peer
	-pkill membersrvc

fabric_reset: fabric_clean docker_chaincode_clean
	rm -rf /var/hyperledger/*
```

Before running a test I typically do a `fabric_reset` that kills all peers and membersrvc, wipes out the blockchain database, kills all Docker processes and deletes all of the chaincode containers.

One of the reasons I hesitated suggesting my own Makefile was because testing consistency requires these Draconian measures, which then requires that  a user be running in a private VM (like Vagrant) or on a private system. I was thinking about how to guarantee that a user understood what havoc they might be wreaking if anyone else was using their machine at the same time.

ghaskins commented:
@bcbrock All excellent points.  Perhaps trying to capture execution environments (such as the unit-test and behave) were a bit overzealous at least in this early pass and I should stick with just dependency/pre-requisite management and let the user run these by hand?  Or should we go the other direction and try to automate the entire thing?  If the former, that simple.  If the latter, perhaps we really should look at more modern tooling (e.g. Gradle) as I know they are often more advanced in this regard.

Regarding the peer target not detecting code changes, this is a good point.  My initial inclination was that could be managed by the user (via a make clean cycle) because detecting all the .go code impacted might be expensive to do each time the user invokes make.  But now that you bring it up, I wonder if we can cache an index so that its only expensive the first time.

ghaskins commented:
@bcbrock another idea: if we switch to "go install" and make the target phony, go can figure it out and detect its own stale objects internally

smithbk commented:
Going forward, we also need to incorporate building and running tests (unit and behave) for each client SDK.   Anya will be doing the node client SDK portion.

ghaskins commented:
@smithbk yeah, i fully expect whatever this mechanism ends up looking like to be extended for cases such as the one you bring up.

ghaskins commented:
@bcbrock You had said: "In yours this is only a side-effect of running behave".  Note that its slightly more granular that this.  There is a "make behave-deps" which only generates the artifacts (if they don't already exist either from previously running behave-deps or the unit-test target).  Aside from the name implying a relationship to behave (which we can generalize), it seems that prereq could be utilized by your workflow too, couldn't it?

bcbrock commented:
I move for automating the whole thing with make targets. I have not seen anything complicated enough yet that cries out for another build system. To me the Makefile is just a repository for simple named scripts.

The .PHONY target should work with Go dependency checking. `go build` is pretty fast when nothing changed.

Yes, I could use behave-deps. I also have  a target `make behave_with_logs`.

If it were me I would also put some brief documentation at the top of the Makefile, like the bulleted list you included at the very top of this note, just so it's all in one place.

ghaskins commented:
@bcbrock seems reasonable to me.  I would prefer to stick with Make as well for the reasons already mentioned (ubiquity, simplicity, and its what I know).

Regarding your comments about phony+go-build and the documentation, those are good, ill add them to the PR.

Regarding "behave-deps" name, any suggestion for a more general name?

Regarding the feature automation, my inclination would be to leave them approximately what I have now and we can enhance their function over time with follow up PRs.

Any objections?

bcbrock commented:
@ghaskins I say `make images`, with options to build only 1 or the other since they take quite a while

jeffgarratt commented:
@ghaskins @jyellick @srderson  If we could get a java JDK installed then we could add Gradle at any point.  I think someone mentioned doing this, not sure if it was in plan or already there.

ghaskins commented:
@jeffgarratt FYI, JRE is available in vagrant/docker as of PR #1250

ghaskins commented:
Replaced by PR #1374

tock-ibm committed:
FAB-17890 Ch.Part.API: allow registrar to list a single channel

Implement the ChannelInfo(..) method in the registrar that looks up
a channel in the chains map and reports extended information on its
status. For this end we have to introduce the follower.Chain and
augment existing cluster type chains with means of reporting their
status.

- Introduce the skeleton a new type of consensus.Chain
  implementation: follower.Chain. This will be created and run when
  the orderer is required to follow the cluster and pull blocks from
  other orderers.

- The plan (for future commits) is for the follower.Chain to trigger
  the creation of an etcdraft.Chain when it discovers the orderer was
  added to the cluster, and vise versa; the etcdraft.Chain will replace
  itself with a follower.Chain when the orderer is removed from the
  cluster.

- Introduce a new interface that cluster-type chains implement, that
  allows them to report their relation to the cluster and their status.
  This is done because the registrar is not aware of the exact type of
  the chains it is keeping. The registrar cannot reflect on the type
  as well, as this will cause an import cycle (due to the etcdraft
  package importing multichannel).

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: Ia454f47f04a8ba3dcd76886a5919d1c734c01015

yacovm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -401,8 +401,20 @@ func (r *Registrar) ChannelList() types.ChannelList {
+		return info, types.ErrChannelNotExist
+	}
+
+	info.Name = channelID
```
can't you initialize them in a single directive?

tock-ibm replied:
I am not sure what you mean... like so:
```go
info.Name, info.Height = channelID, cs.Height()
```

versus

```go
info.Name = channelID
info.Height = cs.Height()
```

yacovm replied:
no, like: info := &InfoObject{ .... }

tock-ibm replied:
there is already an empty info object above that I return in case of error, here I just fill it up... not sure what else I can do. This saves only a few characters anyway.

yacovm commented on orderer/consensus/cluster_status.go:
```diff
@@ -0,0 +1,30 @@
+type StatusReporter interface {
+	// StatusReport provides the cluster relation and status.
+	// See:  channelparticipation.ChannelInfo for more details.
+	StatusReport() (clusterRelation, status string)
```
I am not sure they should be string. Why not enums?

tock-ibm replied:
see below

yacovm commented on orderer/consensus/etcdraft/chain.go:
```diff
@@ -1315,6 +1315,11 @@ func (c *Chain) ValidateConsensusMetadata(oldMetadataBytes, newMetadataBytes []b
 
+// StatusReport returns the ClusterRelation & Status
+func (c *Chain) StatusReport() (string, string) {
+	return "member", "active"
```
Perhaps instead of constructing the API results in a loose manner such as here, we define constants and return them?
That's also why I suggest using enums, so that you can have different types for "relation" and "status".  Also I am not sure why we need 2 different aspects. 
We have "member", and "follower" for a role, and a status.
Can't we flatten them with a Cartesian product of all possible combinations into a single enum? How many are there? I'm sure it's not too many...

tock-ibm replied:
These go directly into the `types.ChannelInfo` struct which gets marshaled into json and sent in the http response. I do not see a point in having an intermediate type as an enum, and then translating to the strings that need to go into the json response.

As you can see in the documentation below, the "ClusterRelation" and "Status"  fields go into separate fields in that struct, which is what the API design calls for. So the internal representation follows the API to avoid unnecessary translations. 

Note also the we have, for backwards compatibility, not just "member" & "follower" but also "config-tracker" & "none". The status is "active", "onboarding" and "inactive". So I wouldn't want to create all the combinations (as these might also evolve in the future).

```
type ChannelInfo struct {
	// The channel name.
	Name string `json:"name"`
	// The channel relative URL (no Host:Port, only path), e.g.: "/participation/v1/channels/my-channel".
	URL string `json:"url"`
	// Whether the orderer is a “member” or ”follower” of the cluster, or "config-tracker" of the cluster, for this channel.
	// For non cluster consensus types (solo, kafka) it is "none".
	// Possible values:  “member”, ”follower”, "config-tracker", "none".
	ClusterRelation string `json:"clusterRelation"`
	// Whether the orderer is ”onboarding”, ”active”, or "inactive", for this channel.
	// For non cluster consensus types (solo, kafka) it is "active".
	// Possible values:  “onboarding”, ”active”, "inactive".
	Status string `json:"status"`
	// Current block height.
	Height uint64 `json:"height"`
}
```

What I can do, and thought of doing but sustained, is to create constants in the types package, e.g.
```go
const (
	ChannelInfoClusterRelationMember string = "member"
)
```
and so on.

yacovm replied:
Constants will do I guess :+1:

tock-ibm replied:
or perhaps something like 

```go

type ClusterRelation string
const (
	ClusterRelationMember ClusterRelation = "member"
)
```

yacovm replied:
Yes this is better

tock-ibm committed:
Review comments: add constants

Add typed constatnts for the possible values of
ChannelInfo.Status & ChannelInfo.ClusterRelation.

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I71062b944d4e5236c4df1462fa73740b68e2065b

yacovm commented on orderer/common/types/channelinfo.go:
```diff
@@ -29,6 +29,23 @@ type ChannelInfoShort struct {
+const (
+	ClusterRelationMember        ClusterRelation = "member"
+	ClusterRelationFollower      ClusterRelation = "follower"
+	ClusterRelationConfigTracker ClusterRelation = "config-tracker"
```
what's a config tracker? I don't see anything about it in the RFC, and what's none for?

tock-ibm replied:
This indeed was not in the RFC because the RFC did not address how this API would behave on a "legacy" system that runs in the system channel model. We have to add it to the RFC some time for completeness. 

The ClusterRelation="config-tracker" is the inactive.Chain. It shows up in the "List" of channels, and we have to differentiate it from the etcdraft.Chain which is "member"/"active". 

The ClusterRelation="none" is for kafka and solo. I know they are deprecated, but they are still in the code-base, so I addressed it.

It is all written on the documentation of the respective ChannelInfo.ClusterRelation field, but I guess it needs to be on the constant as well.

yacovm replied:
Makes sense, thanks

tock-ibm committed:
Review comments: improve documentation

Document types.ClusterRelation and types.Status.

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I8171e6cfe6c3a137e7f1f48a14d1b2fe7f1aea5a
